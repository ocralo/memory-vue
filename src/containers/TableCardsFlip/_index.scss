.table-cards-game {
	@apply flex flex-wrap justify-center gap-4 p-8;

	max-width: 1000px;
	min-width: 290px;
	margin: 0 auto;
	margin-bottom: 2rem;
	background: url('https://img.freepik.com/foto-gratis/pantalla-verde-fondo-verde-que-dice-verde_1340-36533.jpg?w=2000&t=st=1691356976~exp=1691357576~hmac=abca4c003dbe7d8f92605ff27628cef45173c96de63c228dd71bf59f5296bf0c');
	background-size: cover;
	background-position: center;
	border: 5px solid #663300;
	box-shadow: 0 0 20px rgba(0, 0, 0, 0.5);
	border-radius: 10px;

	@media screen and (max-width: 1050px) {
		margin-left: 2rem;
		margin-right: 2rem;
	}

	@media screen and (max-width: 360px) {
		margin-left: 0.5rem;
		margin-right: 0.5rem;
	}
}
