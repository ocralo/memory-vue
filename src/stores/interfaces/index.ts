import type { AnimalsGameClient } from '@/repository/animals/interfaces/client'

export interface AnimalsGameStore extends AnimalsGameClient.Animals {
	flip: boolean
	completed: boolean
}
