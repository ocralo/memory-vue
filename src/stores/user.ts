import { defineStore } from 'pinia'
import { ref } from 'vue'

export const useUserStore = defineStore('user', () => {
	const user = ref<string | null>(null)

	const setUser = (name: string): void => {
		user.value = name
	}

	const clearUser = (): void => {
		user.value = null
	}

	return { user, setUser, clearUser }
})
