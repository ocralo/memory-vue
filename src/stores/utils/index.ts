import type { AnimalsGameStore } from '@/stores/interfaces'

export const duplicateAnimals = (
	arr: AnimalsGameStore[],
): AnimalsGameStore[] => {
	const auxArray = []
	for (let i = 0; i < arr.length; i++) {
		auxArray.push({ ...arr[i], uuid: `${arr[i].uuid}-2` })
	}
	return [...arr, ...auxArray]
}
