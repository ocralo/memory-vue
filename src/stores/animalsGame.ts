import { ref } from 'vue'
import { defineStore } from 'pinia'
import { getAnimals } from '@/repository/animals'
import { PROMISE_STATE } from '@/utils/enums'
import type { AnimalsGameStore } from '@/stores/interfaces'
import { duplicateAnimals } from '@/stores/utils'
import { shuffleArray } from '@/utils'

export const useAnimalGameStore = defineStore('animals', () => {
	const animalsState = ref<AnimalsGameStore[] | null>(null)
	const promiseState = ref<PROMISE_STATE>(PROMISE_STATE.IDLE)
	const errorState = ref<any | null>(null)
	const successPoints = ref<number>(0)
	const errorPoints = ref<number>(0)
	const maxAnimals = ref<number>(4)

	const obtainAnimals = async (): Promise<void> => {
		promiseState.value = PROMISE_STATE.PENDING
		try {
			const response = await getAnimals({ perPage: maxAnimals.value })

			const animalsFlip = response?.animals?.map((animal) => ({
				...animal,
				flip: false,
				completed: false,
			}))

			if (animalsFlip) {
				const animalsDuplicate = duplicateAnimals(animalsFlip)
				const animalsShuffle = shuffleArray(animalsDuplicate)

				animalsState.value = animalsShuffle
				promiseState.value = PROMISE_STATE.RESOLVED
			} else {
				animalsState.value = null
				promiseState.value = PROMISE_STATE.REJECTED
			}
		} catch (error) {
			promiseState.value = PROMISE_STATE.REJECTED
			animalsState.value = null
			errorState.value = error
		}
	}

	const flipAnimalCard = (flag: boolean, id: string): void => {
		if (animalsState.value && Array.isArray(animalsState.value)) {
			const index = animalsState.value.findIndex((animal) => animal.uuid === id)
			if (index >= 0) animalsState.value[index].flip = flag
		}
	}

	const resetGame = (): void => {
		animalsState.value = null
		promiseState.value = PROMISE_STATE.IDLE
		errorState.value = null
		successPoints.value = 0
		errorPoints.value = 0
	}

	const addSuccessPoint = (): void => {
		successPoints.value += 1
	}

	const addErrorPoint = (): void => {
		errorPoints.value += 1
	}

	const completedAnimal = (id: string): void => {
		if (animalsState.value && Array.isArray(animalsState.value)) {
			const index = animalsState.value.findIndex((animal) => animal.uuid === id)
			if (index >= 0) animalsState.value[index].completed = true
		}
	}

	const setMaxAnimals = (max: number): void => {
		maxAnimals.value = max
		resetGame()
	}

	return {
		animalsState,
		maxAnimals,
		errorState,
		promiseState,
		successPoints,
		errorPoints,
		obtainAnimals,
		flipAnimalCard,
		resetGame,
		addSuccessPoint,
		addErrorPoint,
		completedAnimal,
		setMaxAnimals,
	}
})
