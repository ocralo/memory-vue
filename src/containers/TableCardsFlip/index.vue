<template>
	<ModalComponent :isOpen="isOpenModal" @close="handleCloseModal">
		<template v-slot:title>
			<h4 class="text-lg font-medium uppercase mr-4">
				{{ t('modal.congratulations_win') }}
			</h4>
		</template>
		<template v-slot:content>
			<h5 class="text-lg font-medium mb-2 ml-2">
				{{ t('modal.your_score') }}:
			</h5>
			<section class="mb-6 ml-4">
				<h3 class="text-sm text-rose-600 font-medium uppercase mb-1">
					<span>{{ t('modal.error_points') }}:</span> {{ errorPoints }}
				</h3>

				<h3 class="text-sm text-green-600 font-medium uppercase">
					<span>{{ t('modal.success_points') }}:</span> {{ successPoints }}
				</h3>
			</section>
			<section>
				<button
					@click="handleSuccessModal"
					:title="t('modal.close')"
					type="button"
					class="bg-blue-800 text-white px-4 py-2 rounded-md shadow-md hover:bg-blue-900 transition duration-300 ease-in-out w-full">
					{{ t('modal.close') }}
				</button>
			</section>
		</template>
	</ModalComponent>
	<section
		class="w-full h-full grid place-content-center"
		v-if="promiseState === PromisePending">
		<SpinnerComponent />
	</section>
	<template v-else>
		<aside class="flex justify-center gap-4 mb-6">
			<h3 class="text-xl text-rose-600 font-medium uppercase">
				<span>{{ t('tableCardFlip.error_points') }}:</span> {{ errorPoints }}
			</h3>
			<h3 class="text-slate-900 text-xl font-medium">/</h3>
			<h3 class="text-xl text-green-600 font-medium uppercase">
				<span>{{ t('tableCardFlip.success_points') }}:</span>
				{{ successPoints }}
			</h3>
		</aside>
		<main v-if="animals" class="table-cards-game">
			<CardFlip
				v-for="animal in animals"
				@click="handleFlip(animal)"
				:flip="animal.flip"
				:title="animal.image.alt"
				:disabled="animal.completed"
				:back-card="{
					name: animal.image.alt,
					src: animal.image.url,
				}"
				type="button"
				:key="animal.uuid" />
		</main>
	</template>
</template>

<script lang="ts">
	import { defineComponent, ref, watch } from 'vue'
	import { storeToRefs } from 'pinia'
	import { useAnimalGameStore } from '@/stores/animalsGame'
	import SpinnerComponent from '@/components/Spinner/index.vue'

	import CardFlip from '@/components/CardFlip/index.vue'
	import ModalComponent from '@/components/Modal/index.vue'
	import type { AnimalsGameStore } from '@/stores/interfaces'
	import { useRouter } from 'vue-router'
	import { RoutePath } from '@/router/enum'
	import { useUserStore } from '@/stores/user'

	export default defineComponent({
		name: 'TableCardsFlip',
	})
</script>

<script setup lang="ts">
	import { useI18n } from 'vue-i18n'
	import { PROMISE_STATE } from '@/utils/enums'

	const { t } = useI18n()
	const router = useRouter()

	const PromisePending = PROMISE_STATE.PENDING

	const userStore = useUserStore()
	const { clearUser } = userStore

	const animalStore = useAnimalGameStore()
	const { animalsState, promiseState, errorPoints, successPoints, maxAnimals } =
		storeToRefs(animalStore)
	const {
		obtainAnimals,
		flipAnimalCard,
		addSuccessPoint,
		addErrorPoint,
		completedAnimal,
		resetGame,
	} = animalStore

	const animals = ref(animalsState || null)

	const flippedCards = ref<string[]>([])
	const isOpenModal = ref(false)

	const handleFlip = (animal: AnimalsGameStore) => {
		if (!animal.flip) {
			flippedCards.value.push(animal.uuid)
		} else {
			flippedCards.value = flippedCards.value.filter(
				(uuid) => uuid !== animal.uuid,
			)
		}
		flipAnimalCard(!animal.flip, animal.uuid)
	}

	const restartTwoCards = (
		firstCard: AnimalsGameStore,
		secondCard: AnimalsGameStore,
	) => {
		setTimeout(() => {
			flipAnimalCard(false, firstCard?.uuid)
			flipAnimalCard(false, secondCard?.uuid)
		}, 1000)
	}

	const completedPairs = (uuid: string[]) => {
		for (const id of uuid) {
			completedAnimal(id)
		}
	}

	const flippedTwoCards = (value: string[]) => {
		if (value.length === 2) {
			const [firstCard, secondCard] = value
			const firstAnimal = animals.value?.find(
				(animal) => animal.uuid === firstCard,
			)
			const secondAnimal = animals.value?.find(
				(animal) => animal.uuid === secondCard,
			)
			if (firstAnimal?.slug === secondAnimal?.slug) {
				completedPairs(value)
				addSuccessPoint()
			} else {
				if (firstAnimal?.uuid && secondAnimal?.uuid) {
					restartTwoCards(firstAnimal, secondAnimal)
				}
				addErrorPoint()
			}
			flippedCards.value = []
		}
	}

	const handleCloseModal = () => {
		isOpenModal.value = false
	}

	const handleSuccessModal = () => {
		handleCloseModal()
		clearUser()
		resetGame()
		router.push(RoutePath.Home)
	}

	watch(
		flippedCards,
		(newValue) => {
			flippedTwoCards(newValue)
		},
		{ deep: true },
	)

	watch(successPoints, (newValue) => {
		if (newValue >= maxAnimals.value) {
			isOpenModal.value = true
		}
	})

	obtainAnimals()
</script>

<style src="./_index.scss" scoped />
