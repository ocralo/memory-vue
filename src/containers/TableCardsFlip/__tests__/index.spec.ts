import { describe, it, expect, beforeEach, afterEach } from 'vitest'
import { createTestingPinia } from '@pinia/testing'
import { shallowMount } from '@vue/test-utils'
import sinon from 'sinon'

import { useAnimalGameStore } from '@/stores/animalsGame'
import TableCardsFlip from '@/containers/TableCardsFlip/index.vue'

describe('TableCardsFlip', () => {
	let wrapper: any
	let store: any

	beforeEach(() => {
		wrapper = shallowMount(TableCardsFlip, {
			global: {
				plugins: [createTestingPinia({ createSpy: sinon.spy })],
			},
		})

		store = useAnimalGameStore()
	})

	afterEach(() => {
		wrapper.unmount()
	})

	describe('mount', () => {
		it('should render correctly', () => {
			expect(wrapper.html()).toMatchSnapshot()
		})
	})

	describe('methods', () => {
		describe('handleFlip', () => {
			it('should call flipAnimalCard', () => {
				wrapper.vm.handleFlip(1)
				expect(store.flipAnimalCard.calledOnce).toBeTruthy()
			})
		})

		describe('completedPairs', () => {
			it('should call completedPairs', () => {
				wrapper.vm.completedPairs(['1', '2'])
				expect(store.completedAnimal.called).toBeTruthy()
			})
		})

		describe('handleCloseModal', () => {
			it('should call handleCloseModal', () => {
				wrapper.vm.handleCloseModal()
				expect(wrapper.vm.isOpenModal).toBeFalsy()
			})
		})
	})
})
