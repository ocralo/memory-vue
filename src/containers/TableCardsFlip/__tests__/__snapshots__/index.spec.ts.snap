// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TableCardsFlip > mount > should render correctly 1`] = `
"<modal-component-stub data-v-978a112d=\\"\\" isopen=\\"false\\"></modal-component-stub>
<aside data-v-978a112d=\\"\\" class=\\"flex justify-center gap-4 mb-6\\">
  <h3 data-v-978a112d=\\"\\" class=\\"text-xl text-rose-600 font-medium uppercase\\"><span data-v-978a112d=\\"\\">tableCardFlip.error_points:</span> 0</h3>
  <h3 data-v-978a112d=\\"\\" class=\\"text-slate-900 text-xl font-medium\\">/</h3>
  <h3 data-v-978a112d=\\"\\" class=\\"text-xl text-green-600 font-medium uppercase\\"><span data-v-978a112d=\\"\\">tableCardFlip.success_points:</span> 0</h3>
</aside>
<!--v-if-->"
`;
